Why do we use Relay? (#553)
artsy/artsy.github.io: _posts/2019-04-10-omakase-relay.markdown

simplified [profile page] from the Artsy iOS app:

```ts
import React from "react"
import { createFragmentContainer, graphql } from "react-relay"
import { MyProfile_me } from "__generated__/MyProfile_me.graphql"

interface Props extends ViewProperties {
  me: MyProfile_me
}

export class MyProfile extends React.Component<Props> {
  render() {
    return (
      <View>
        <Header>
          <ProfilePhoto initials={this.props.me.initials} image={this.props.me.image} />
          <Subheading>{this.props.me.name}</Subheading>
        </Header>
        <ButtonSection>
          <ProfileButton
            section="Selling"
            description="Sell works from your collection"
            onPress={startSubmission}
          />
          <ProfileButton
            section="Account Details"
            description="Email, password reset, profile"
            onPress={goToUserSettings}
          />
        </ButtonSection>
      </View>
    )
  }
}

export default createFragmentContainer(MyProfile, {
  me: graphql`
    fragment MyProfile_me on Me {
      name
      image
      initials
    }
  `
})
```

There are three moving parts:

> - The TypeScript interface `MyProfile_me` which ensures we can only use fields that were selected in the fragment

Reviewer: ```suggestion
- The TypeScript interface `MyProfile_me`, generated by the compiler, which ensures we can only use fields that were selected in the fragment
```

passed through the React tree. E.g.

{% include epic_img.html url="/images/omakase-relay/tree.png" title="REST inspired props" style="width:100%;" %}

This means most components know more about the request than it probably needs, as it may be needed to pass on to
the component's children. This can lead to over-fetching, or even worse, not knowing if you can delete or refactor
a component.

Data masking solves this by hiding data that the component didn't request. I've still yet to find the right visual
abstraction, but I feel this just about pays for itself.

{% include epic_img.html url="/images/omakase-relay/isolation.png" title="Relay isolation tree" style="width:100%;" %}

You let Relay be responsible for consolidating all your fragments into a query via the `QueryRenderer`, causing the
network request, and your response data to be passed your through your component hierarchy. This means Relay
powered components can be safely changed and drastically reduces the chance for unintended consequences elsewhere.

This isolation gives Artsy engineers the safety to work on projects with tens of contributors, which change over

> long time periods without accruing technical debt. The components we create are nearly all focused only on the

Reviewer: ```suggestion
time, need to onboard onto a project easily, and only work on isolated parts of the codebase without accruing technical debt. The components we create are nearly all focused only on the
```

Whaddayathink?

GraphQL schema, it emits Flow types for your fragment’s field selections–which we’ve extended to emit TypeScript
types instead, and there are strict naming systems enforced by the compiler. All of these help guide engineers to
build scalable codebases.

How this works in practice is that whenever you need to change the data a component requires, you edit the
fragment, the Relay compiler verifies your query, if successful then your TypeScript types are updated and you can
use the new property in your React component above. See below for a [quick video][vid] showing the Relay compiler
in action:

{% include epic_video.html url="/images/omakase-relay/relay-process-720.mov" title="Relay isolation tree" style="width:100%; display:block;" %}

Moving all of these checks to be during dev-time means we can feel more confident in our deploys. This is
especially an issue in an iOS native codebase, when any deploy requires a review from Apple and roll-backs are
impossible.

## Cultural Fit

Relay fit well into our team because:

- We had engineers who were interested in contributing back and extending Relay to work for our cases

> - We had engineers that were used to the benefits of ahead-of-time error validation tools like compilers

Reviewer: ```suggestion
- We had engineers that were used to the benefits of ahead-of-time error validation tools that compilers offer
```
